core/connection/manager.go

```diff
@@ -159,39 +137,83 @@ func (manager *connectionManager) startConnection(consumerID identity.Identity,
 		return err
 	}
 
-	sessionCreateConfig, err := connection.GetConfig()
+	sessionDTO, paymentInfo, err := manager.createSession(connection, dialog, consumerID, proposal)
 	if err != nil {
 		return err
 	}
 
+	err = manager.launchPayments(paymentInfo, dialog, consumerID, providerID)
+	if err != nil {
+		return err
+	}
+
+	err = manager.startConnection(connection, consumerID, proposal, params, sessionDTO, stateChannel, statisticsChannel)
+	if err == context.Canceled {
+		return ErrConnectionCancelled
+	}
+	return err
+}
+
+func (manager *connectionManager) launchPayments(paymentInfo *session.PaymentInfo, dialog communication.Dialog, consumerID, providerID identity.Identity) error {
+	var promiseState promise.State
+	if paymentInfo != nil {
+		promiseState.Amount = paymentInfo.LastPromise.Amount
+		promiseState.Seq = paymentInfo.LastPromise.SequenceID
+	}
+
 	messageChan := make(chan balance.Message, 1)
 
-	// TODO: load initial promise state
-	payments, err := manager.paymentIssuerFactory(promise.State{}, messageChan, dialog, consumerID, providerID)
+	payments, err := manager.paymentIssuerFactory(promiseState, messageChan, dialog, consumerID, providerID)
 	if err != nil {
 		return err
 	}
 
-	cancel = append(cancel, func() { payments.Stop() })
+	manager.cleanup = append(manager.cleanup, func() { payments.Stop() })
```

Comment:
should be enough:
```suggestion
	manager.cleanup = append(manager.cleanup, payments.Stop)
```
We should write function wrapper only in cases we have not a real function but some channel. There are more such places with append.

Reply:
fixed this one, all the other places do need a wrapper since we're using `func()` as the array type, and all the others tend to return the error. We could switch this to using the `func() error` for the array type, but I don't think we'd save much of anything

Reply:
but thats a bit smelly then :) we might be swallowing errors. Maybe we can collect '[]func() error' then and print out errors when interating through cleanup? Would be bonus not having those nasty wrappers and all error accounted.